dll/win32/shell32/shlexec.cpp

```diff
@@ -1983,21 +1983,49 @@ static BOOL SHELL_execute(LPSHELLEXECUTEINFOW sei, SHELL_ExecuteW32 execfunc)
     else
         *wszParameters = L'\0';
 
+    // Get the working directory
     WCHAR dirBuffer[MAX_PATH];
     LPWSTR wszDir = dirBuffer;
+    wszDir[0] = UNICODE_NULL;
     CHeapPtr<WCHAR, CLocalAllocator> wszDirAlloc;
-    if (sei_tmp.lpDirectory)
+    if (sei_tmp.lpDirectory && *sei_tmp.lpDirectory)
     {
-        len = lstrlenW(sei_tmp.lpDirectory) + 1;
-        if (len > _countof(dirBuffer))
+        if (sei_tmp.fMask & SEE_MASK_DOENVSUBST)
         {
-            wszDirAlloc.Allocate(len);
-            wszDir = wszDirAlloc;
+            LPWSTR tmp = expand_environment(sei_tmp.lpDirectory);
+            if (tmp)
+            {
+                wszDirAlloc.Attach(tmp);
+                wszDir = wszDirAlloc;
+            }
+        }
+        else
+        {
+            __SHCloneStrW(&wszDirAlloc, sei_tmp.lpDirectory);
+            if (wszDirAlloc)
+                wszDir = wszDirAlloc;
+        }
+    }
+    if (!wszDir[0])
+    {
+        ::GetCurrentDirectoryW(_countof(dirBuffer), dirBuffer);
+        wszDir = dirBuffer;
+    }
+    // NOTE: ShellExecute should accept the invalid working directory for historical reason.
+    // NOTE: ".\\" is a special case.
+    if (!PathIsDirectoryW(wszDir) || lstrcmpiW(wszDir, L".\\") == 0)
```

Comment:
You claim that ".\\" is a special case. But what about "..\\" ? And does this need the trailing slash? What if the path is something like ".\\..\\" etc?

Reply:
> what about "..\" ?

Nope.

> does this need the trailing slash?

Yep.

> What if the path is something like ".\..\" etc?

Nope.

Reply:
Deleted the special case.